src/bodygraph.cpp

```diff
@@ -0,0 +1,567 @@
+    std::string selected_sym;
+    if( !!selected_graph ) {
+        for( const auto &bgp : id->parts ) {
+            if( selected_graph == &bgp.second ) {
+                selected_sym = bgp.first;
+            }
+        }
+    }
+    const bool hflip = !!id->mirror;
+    const bodygraph_id &rid = get_bg_rows( id );
+    for( unsigned i = 0; i < rid->rows.size() && i < BPGRAPH_MAXROWS; i++ ) {
+        int j = hflip ? rid->rows[i].size() - 1 : 0;
+        for( int x = 0 ; x < BPGRAPH_MAXCOLS && j < BPGRAPH_MAXCOLS && j >= 0; hflip ? j-- : j++, x++ ) {
+            std::string sym = id->fill_sym.empty() ? rid->rows[i][j] : id->fill_sym;
+            nc_color col = id->fill_color;
+            auto iter = id->parts.find( rid->rows[i][j] );
+            if( iter != id->parts.end() ) {
+                sym = iter->second.sym;
+            }
+            if( rid->rows[i][j] == " " ) {
+                col = c_unset;
+                sym = " ";
+            } else if( rid->rows[i][j] == selected_sym ) {
+                col = id->parts.at( selected_sym ).sel_color;
+            }
+            mvwputch( w_graph, point( x, i ), col, sym );
+        }
+    }
+    wnoutrefresh( w_graph );
+}
+
+void bodygraph_display::draw_info()
+{
+    werase( w_info );
+    int y = 0;
+    for( unsigned i = top_info; i < info_txt.size() && y < BPGRAPH_HEIGHT - 2; i++, y++ ) {
+        if( info_txt[i] == "--" ) {
+            for( int x = 1; x < info_width - 2; x++ ) {
+                mvwputch( w_info, point( x, y ), c_dark_gray, LINE_OXOX );
+            }
+        } else {
+            trim_and_print( w_info, point( 1, y ), info_width - 2, c_white, info_txt[i] );
+        }
+    }
+    wnoutrefresh( w_info );
+}
+
+void bodygraph_display::prepare_partlist()
+{
+    partlist.clear();
+    for( const auto &bgp : id->parts ) {
+        for( const bodypart_id &bid : bgp.second.bodyparts ) {
+            partlist.emplace_back( std::make_tuple( bid, static_cast<const sub_body_part_type *>( nullptr ),
+                                                    &bgp.second ) );
+        }
+        for( const sub_bodypart_id &sid : bgp.second.sub_bodyparts ) {
+            partlist.emplace_back( std::make_tuple( sid->parent.id(), &*sid, &bgp.second ) );
+        }
+    }
+    std::sort( partlist.begin(), partlist.end(),
+               []( const std::tuple<bodypart_id, const sub_body_part_type *, const bodygraph_part *> &a,
+    const std::tuple<bodypart_id, const sub_body_part_type *, const bodygraph_part *> &b ) {
+        if( !!std::get<1>( a ) && !!std::get<1>( b ) ) {
+            return std::get<1>( a )->name.translated_lt( std::get<1>( b )->name );
+        }
+        if( std::get<0>( a )->name.translated_ne( std::get<0>( b )->name ) ) {
+            return std::get<0>( a )->name.translated_lt( std::get<0>( b )->name );
+        }
+        if( !std::get<1>( a ) != !std::get<1>( b ) ) {
+            return !std::get<1>( a );
+        }
+        return true;
```

Comment:
```suggestion
        return a != b;
```
to satisfy this requirement
`For all a, comp(a,a)==false`
and fix this asan error:
<details>
<summary>asan.log</summary>

```gdb
==1308027==ERROR: AddressSanitizer: container-overflow on address 0x613000f2e0f8 at pc 0x5582eab4374c bp 0x7ffe8da2c230 sp 0x7ffe8da2c228
READ of size 8 at 0x613000f2e0f8 thread T0
    #0 0x5582eab4374b in bodygraph_display::prepare_partlist()::$_2::operator()(std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*> const&, std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*> const&) const /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/bodygraph.cpp:394:15
    #1 0x5582eab42a10 in void std::__1::__sort<bodygraph_display::prepare_partlist()::$_2&, std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*>*>(std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*>*, std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*>*, bodygraph_display::prepare_partlist()::$_2&) /usr/bin/../include/c++/v1/__algorithm/sort.h:403:24
    #2 0x5582eab3d4ae in void std::__1::sort<std::__1::__wrap_iter<std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*>*>, bodygraph_display::prepare_partlist()::$_2>(std::__1::__wrap_iter<std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*>*>, std::__1::__wrap_iter<std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*>*>, bodygraph_display::prepare_partlist()::$_2) /usr/bin/../include/c++/v1/__algorithm/sort.h:514:9
    #3 0x5582eab3d4ae in bodygraph_display::prepare_partlist() /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/bodygraph.cpp:391:5
    #4 0x5582eab41545 in bodygraph_display::display() /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/bodygraph.cpp:516:5
    #5 0x5582eab42357 in display_bodygraph(string_id<bodygraph> const&) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/bodygraph.cpp:566:9
    #6 0x5582ecf5e38d in handle_player_display_action(Character&, unsigned int&, unsigned int&, (anonymous namespace)::player_display_tab&, input_context&, ui_adaptor const&, ui_adaptor const&, ui_adaptor const&, ui_adaptor const&, ui_adaptor const&, ui_adaptor const&, ui_adaptor const&, ui_adaptor const&, ui_adaptor const&, std::__1::vector<string_id<mutation_branch>, std::__1::allocator<string_id<mutation_branch> > > const&, std::__1::vector<bionic, std::__1::allocator<bionic> > const&, std::__1::vector<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > > const&, std::__1::vector<HeaderSkill, std::__1::allocator<HeaderSkill> > const&, bool) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/player_display.cpp:1131:9
    #7 0x5582ecf59d61 in Character::disp_info(bool) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/player_display.cpp:1598:16
    #8 0x5582ea8bcd24 in game::do_regular_action(action_id&, avatar&, cata::optional<tripoint> const&) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/handle_action.cpp:2472:30
    #9 0x5582ea8eb2b2 in game::handle_action() /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/handle_action.cpp:2893:14
    #10 0x5582e9c908cb in do_turn() /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/do_turn.cpp:702:24
    #11 0x5582e8be1404 in main /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/main.cpp:766:17
    #12 0x7f15a532330f in __libc_start_call_main libc-start.c
    #13 0x7f15a53233c0 in __libc_start_main@GLIBC_2.2.5 (/usr/lib/libc.so.6+0x2d3c0)
    #14 0x5582e8aeb024 in _start (/usr/src/cataclysm-dda-git/src/build_asan/cataclysm-tiles+0x7471024)

0x613000f2e0f8 is located 248 bytes inside of 384-byte region [0x613000f2e000,0x613000f2e180)
allocated by thread T0 here:
    #0 0x5582e8bdd4f1 in operator new(unsigned long) (/usr/src/cataclysm-dda-git/src/build_asan/cataclysm-tiles+0x75634f1)
    #1 0x5582eab6a3a6 in void* std::__1::__libcpp_operator_new<unsigned long>(unsigned long) /usr/bin/../include/c++/v1/new:235:10
    #2 0x5582eab6a3a6 in std::__1::__libcpp_allocate(unsigned long, unsigned long) /usr/bin/../include/c++/v1/new:261:10
    #3 0x5582eab6a3a6 in std::__1::allocator<std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*> >::allocate(unsigned long) /usr/bin/../include/c++/v1/__memory/allocator.h:108:38
    #4 0x5582eab6a3a6 in std::__1::allocator_traits<std::__1::allocator<std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*> > >::allocate(std::__1::allocator<std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*> >&, unsigned long) /usr/bin/../include/c++/v1/__memory/allocator_traits.h:262:20
    #5 0x5582eab6a3a6 in std::__1::__split_buffer<std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*>, std::__1::allocator<std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*> >&>::__split_buffer(unsigned long, unsigned long, std::__1::allocator<std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*> >&) /usr/bin/../include/c++/v1/__split_buffer:315:29
    #6 0x5582eab69a7f in void std::__1::vector<std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*>, std::__1::allocator<std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*> > >::__emplace_back_slow_path<std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*> >(std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*>&&) /usr/bin/../include/c++/v1/vector:1687:49
    #7 0x5582eab3d393 in bodygraph_display::prepare_partlist() /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/bodygraph.cpp:384:22
    #8 0x5582eab41545 in bodygraph_display::display() /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/bodygraph.cpp:516:5
    #9 0x5582eab42357 in display_bodygraph(string_id<bodygraph> const&) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/bodygraph.cpp:566:9
    #10 0x5582ecf5e38d in handle_player_display_action(Character&, unsigned int&, unsigned int&, (anonymous namespace)::player_display_tab&, input_context&, ui_adaptor const&, ui_adaptor const&, ui_adaptor const&, ui_adaptor const&, ui_adaptor const&, ui_adaptor const&, ui_adaptor const&, ui_adaptor const&, ui_adaptor const&, std::__1::vector<string_id<mutation_branch>, std::__1::allocator<string_id<mutation_branch> > > const&, std::__1::vector<bionic, std::__1::allocator<bionic> > const&, std::__1::vector<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > > const&, std::__1::vector<HeaderSkill, std::__1::allocator<HeaderSkill> > const&, bool) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/player_display.cpp:1131:9
    #11 0x5582ecf59d61 in Character::disp_info(bool) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/player_display.cpp:1598:16
    #12 0x5582ea8bcd24 in game::do_regular_action(action_id&, avatar&, cata::optional<tripoint> const&) /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/handle_action.cpp:2472:30
    #13 0x5582ea8eb2b2 in game::handle_action() /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/handle_action.cpp:2893:14
    #14 0x5582e9c908cb in do_turn() /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/do_turn.cpp:702:24
    #15 0x5582e8be1404 in main /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/main.cpp:766:17
    #16 0x7f15a532330f in __libc_start_call_main libc-start.c

HINT: if you don't care about these errors you may set ASAN_OPTIONS=detect_container_overflow=0.
If you suspect a false positive see also: https://github.com/google/sanitizers/wiki/AddressSanitizerContainerOverflow.
SUMMARY: AddressSanitizer: container-overflow /usr/src/cataclysm-dda-git/src/build_asan/../cataclysm-dda-git/src/bodygraph.cpp:394:15 in bodygraph_display::prepare_partlist()::$_2::operator()(std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*> const&, std::__1::tuple<int_id<body_part_type>, sub_body_part_type const*, bodygraph_part const*> const&) const
Shadow bytes around the buggy address:
  0x0c26801ddbc0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c26801ddbd0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c26801ddbe0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c26801ddbf0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c26801ddc00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x0c26801ddc10: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 fc[fc]
  0x0c26801ddc20: fc fc fc fc fc fc fc fc fc fc fc fc fc fc fc fc
  0x0c26801ddc30: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c26801ddc40: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c26801ddc50: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c26801ddc60: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==1308027==ABORTING
```
</details>

Reply:
Good find, appreciated!